### PR TITLE
Support asset transfers via XCM

### PR DESCRIPTION
**Pull request by vgeddes**
Adds support for the following features:
* Upward/Downward transfer of DOT.
* Transferring bridged Ethereum assets between sibling parachains.

Both kinds of transfers are performed using the reserve-backed model as described in [Polkadot Cross-Consensus Message (XCM) Format](https://github.com/paritytech/xcm-format/blob/master/README.md).

Changes:
* Added a pallet `token-dealer`. This module includes dispatchables for initiating transfers:
  * `transfer_dot_to_relaychain`
  * `transfer_bridged_asset_to_parachain`
* Added a crate `xcm-support` which contains an implementation for XCM `TransactAsset` trait. This is key to enabling deposits and withdrawals of assets by the XCM executor.
* Iterative improvements to the bridged assets pallet.
* Wiring all the above into the runtime,

There is pretty good unit test coverage for all of this in the the `token-dealer` module:
```
vincent@crake:~/code/polkadot-ethereum/parachain/pallets/token-dealer (rococo-v1)$ cargo test
   Compiling artemis-xcm-support v0.1.0 (/home/vincent/code/polkadot-ethereum/parachain/primitives/xcm-support)
   Compiling artemis-token-dealer v0.1.1 (/home/vincent/code/polkadot-ethereum/parachain/pallets/token-dealer)
    Finished test [unoptimized + debuginfo] target(s) in 8.58s
     Running /home/vincent/code/polkadot-ethereum/parachain/target/debug/deps/artemis_token_dealer-10cd52ae3eab93d4

running 3 tests
test tests::transfer_foreign_eth_to_parachain ... ok
test tests::transfer_dot_to_relay_chain ... ok
test tests::transfer_native_eth_to_parachain ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests artemis-token-dealer

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

### CODE REVIEW

**Comment by vgeddes** [comment]
Note: Leaving this PR in draft mode until upstream dependencies have stabilized (i.e. the features we need have been merged into the master branches of Substrate, Polkadot, and Cumulus).

**Review by philipstanislaus — COMMENTED** [review]
LGTM, just a few comments 🎉

File: parachain/node/Cargo.toml [hunk 1]
```diff
@@ -1,145 +1,80 @@
...
+sp-core = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-inherents = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-consensus = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-session = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-offchain = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-block-builder = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-blockchain = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-timestamp = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-trie = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
+sp-transaction-pool = { git = "https://github.com/paritytech/substrate", branch = "rococo-branch" }
```
[the remark this answers is not in the document]
Comment (musnit) [reply parachain/node/Cargo.toml:53]: don't think we should worry about this for now as dependencies are still so much in flux and breaking our code is fine as we'll need to update it anyway soon once they mature

File: parachain/node/src/command.rs [hunk 2]
```diff
@@ -30,40 +60,316 @@ impl SubstrateCli for Cli {
...
+		parachain-collator [parachain-args] -- [relaychain-args]"
+			.into()
+	}
+
+	fn author() -> String {
+		env!("CARGO_PKG_AUTHORS").into()
+	}
+
+	fn support_url() -> String {
+		"https://github.com/paritytech/cumulus/issues/new".into()
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply parachain/node/src/command.rs]: Fixed

File: parachain/node/src/command.rs [hunk 3]
```diff
@@ -30,40 +60,316 @@ impl SubstrateCli for Cli {
...
+	fn impl_version() -> String {
+		env!("SUBSTRATE_CLI_IMPL_VERSION").into()
+	}
+
+	fn description() -> String {
+		"Parachain Collator Template\n\nThe command-line arguments provided first will be \
+		passed to the parachain node, while the arguments provided after -- will be passed \
+		to the relaychain node.\n\n\
+		parachain-collator [parachain-args] -- [relaychain-args]"
+			.into()
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply parachain/node/src/command.rs:87]: Fixed

File: parachain/pallets/assets/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
+//! # Asset
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply parachain/pallets/assets/src/lib.rs]: Nice catch! Fixed.

File: parachain/node/src/command.rs [hunk 5]
```diff
@@ -15,7 +39,13 @@ impl SubstrateCli for Cli {
...
 
 	fn description() -> String {
-		env!("CARGO_PKG_DESCRIPTION").into()
+		format!(
+			"Parachain Collator Template\n\nThe command-line arguments provided first will be \
+		passed to the parachain node, while the arguments provided after -- will be passed \
+		to the relaychain node.\n\n\
+		{} [parachain-args] -- [relaychain-args]",
+			Self::executable_name()
+		)
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply parachain/node/src/command.rs:48]: Yeah, These changes to the files under `parachain/node/src/` were necessary to support the CLI interfaces/abstractions provided by the Polkadot in the `gav-xcmp` branch.

File: parachain/pallets/assets/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+decl_module! {
+
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Transfer some free balance to another account.
+		#[weight = 10]
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply parachain/pallets/assets/src/lib.rs:98]: Done, added https://github.com/Snowfork/polkadot-ethereum/issues/161

Calculating & verifying weights for all the parachain dispatchables is on our roadmap, so we'll get it to it eventually.

**Comment by musnit** [comment]
@vgeddes can close?

**Comment by vgeddes** [comment]
Obsoleted by #200

**Pull request closed without merging**